linera-views/src/views/bucket_queue_view.rs

```diff
@@ -0,0 +1,656 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(with_metrics)]
+use std::sync::LazyLock;
+use std::{
+    collections::{vec_deque::IterMut, VecDeque},
+    fmt::Debug,
+};
+
+use async_trait::async_trait;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+#[cfg(with_metrics)]
+use {
+    linera_base::prometheus_util::{self, MeasureLatency},
+    prometheus::HistogramVec,
+};
+
+use crate::{
+    batch::Batch,
+    common::{from_bytes_option, from_bytes_option_or_default, HasherOutput, MIN_VIEW_TAG},
+    context::Context,
+    hashable_wrapper::WrappedHashableContainerView,
+    views::{ClonableView, HashableView, Hasher, View, ViewError},
+};
+
+#[cfg(with_metrics)]
+/// The runtime of hash computation
+static BUCKET_QUEUE_VIEW_HASH_RUNTIME: LazyLock<HistogramVec> = LazyLock::new(|| {
+    prometheus_util::register_histogram_vec(
+        "bucket_queue_view_hash_runtime",
+        "BucketQueueView hash runtime",
+        &[],
+        Some(vec![
+            0.001, 0.003, 0.01, 0.03, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.0, 5.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+/// Key tags to create the sub-keys of a BucketQueueView on top of the base key.
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the front of the view
+    Front = MIN_VIEW_TAG,
+    /// Prefix for the storing of the stored-indices information
+    Store,
+    /// Prefix for the indices of the log.
+    Index,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+struct StoredIndices {
+    indices: Vec<(usize, usize)>,
+    position: usize,
+}
```

Comment:
Please document this: what indices are these, and the position of what in what?

Looking at the code: Is the first `usize` the number of entries in a bucket, and the second one the index of the first item in that bucket?

Wouldn't it be easier to make all buckets the same size, i.e. if we add elements we add them to the last bucket until it's full? Then we wouldn't need a list of indices at all?

Reply:
The list of indices cannot be all of the same size. Suppose it that way and work out the argument:
* Suppose for the sake of argument that N=5.
* Suppose that the list has 13 elements. But supposed that we have just read the front (of size 5).
* Suppose that we have 2 additional elements in `new_back_values` and that we want to write them in the flush.
* Then the flush operation needs to create a list of 5 elements with 3 from the storage and 2 from the `new_back_values`. 
* So, the flush needs to access the database, which is not possible since flush is not async.

So, yes, the code tries to have the buckets all the same size, but that is not always possible.